### ISSUE TITLE
Issue when using nats as a registry

### ISSUE THREAD

**Issue by alex-godwin** [micro/micro#493]
If you use the image at `micro/micro:latest` and either set the environment variable `MICRO_REGISTRY=nats` or use `--registry=nats` micro will print the default help page and no error like this: 

```
> docker run --name=test micro/micro:latest --registry=nats

NAME:
   micro - A microservice runtime

USAGE:
   micro [global options] command [command options] [arguments...]

VERSION:
   v1.18.0-6b64c1e-1578946501

COMMANDS:
     api         Run the api gateway
     bot         Run the chatops bot
     cli         Run the interactive CLI
     call        Call a service e.g micro call greeter Say.Hello '{"name": "John"}
     services    List the services in the network
     stream      Create a service stream
     publish     Publish a message to a topic
     stats       Query the stats of a service
     list        List items in registry or network
     register    Register an item in the registry
     deregister  Deregister an item in the registry
     get         Get item from registry
     broker      Run the message broker
     health      Check the health of a service
     proxy       Run the service proxy
     monitor     Run the monitoring service
     router      Run the micro network router
     tunnel      Run the micro network tunnel
     network     Run the micro network node
     registry    Run the service registry
     runtime     Run the micro runtime
     run         Run a service e.g micro run service version
     kill        Kill removes a running service e.g micro kill service
     ps          Ps returns status of a running service or lists all running services e.g. micro ps
     debug       Run the micro debug service
     log         Get logs for a service
     server      Run the micro network server
     service     Run a micro service
     store       Run the micro store service
     token       API token commands
     new         Create a service template
     plugin      Plugin commands
     web         Run the web dashboard
     init        Run the micro operator
     help, h     Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --client value                  Client for go-micro; rpc [$MICRO_CLIENT]
   --client_request_timeout value  Sets the client request timeout. e.g 500ms, 5s, 1m. Default: 5s [$MICRO_CLIENT_REQUEST_TIMEOUT]
   --client_retries value          Sets the client retries. Default: 1 (default: 1) [$MICRO_CLIENT_RETRIES]
   --client_pool_size value        Sets the client connection pool size. Default: 1 (default: 0) [$MICRO_CLIENT_POOL_SIZE]
   --client_pool_ttl value         Sets the client connection pool ttl. e.g 500ms, 5s, 1m. Default: 1m [$MICRO_CLIENT_POOL_TTL]
   --register_ttl value            Register TTL in seconds (default: 60) [$MICRO_REGISTER_TTL]
   --register_interval value       Register interval in seconds (default: 30) [$MICRO_REGISTER_INTERVAL]
   --server value                  Server for go-micro; rpc [$MICRO_SERVER]
   --server_name value             Name of the server. go.micro.srv.example [$MICRO_SERVER_NAME]
   --server_version value          Version of the server. 1.1.0 [$MICRO_SERVER_VERSION]
   --server_id value               Id of the server. Auto-generated if not specified [$MICRO_SERVER_ID]
   --server_address value          Bind address for the server. 127.0.0.1:8080 [$MICRO_SERVER_ADDRESS]
   --server_advertise value        Used instead of the server_address when registering with discovery. 127.0.0.1:8080 [$MICRO_SERVER_ADVERTISE]
   --server_metadata value         A list of key-value pairs defining metadata. version=1.0.0 [$MICRO_SERVER_METADATA]
   --broker value                  Broker for pub/sub. http, nats, rabbitmq [$MICRO_BROKER]
   --broker_address value          Comma-separated list of broker addresses [$MICRO_BROKER_ADDRESS]
   --profile value                 Debug profiler for cpu and memory stats [$MICRO_DEBUG_PROFILE]
   --registry value                Registry for discovery. etcd, mdns [$MICRO_REGISTRY]
   --registry_address value        Comma-separated list of registry addresses [$MICRO_REGISTRY_ADDRESS]
   --runtime value                 Runtime for building and running services e.g local, kubernetes (default: "local") [$MICRO_RUNTIME]
   --selector value                Selector used to pick nodes for querying [$MICRO_SELECTOR]
   --store value                   Store used for key-value storage [$MICRO_STORE]
   --store_address value           Comma-separated list of store addresses [$MICRO_STORE_ADDRESS]
   --store_namespace value         Namespace for store data [$MICRO_STORE_NAMESPACE]
   --transport value               Transport mechanism used; http [$MICRO_TRANSPORT]
   --transport_address value       Comma-separated list of transport addresses [$MICRO_TRANSPORT_ADDRESS]
   --local                         Enable local only development
   --enable_acme                   Enables ACME support via Let's Encrypt. ACME hosts should also be specified. [$MICRO_ENABLE_ACME]
   --acme_hosts value              Comma separated list of hostnames to manage ACME certs for [$MICRO_ACME_HOSTS]
   --acme_provider value           The provider that will be used to communicate with Let's Encrypt. Valid options: autocert, certmagic [$MICRO_ACME_PROVIDER]
   --enable_tls                    Enable TLS support. Expects cert and key file to be specified [$MICRO_ENABLE_TLS]
   --tls_cert_file value           Path to the TLS Certificate file [$MICRO_TLS_CERT_FILE]
   --tls_key_file value            Path to the TLS Key file [$MICRO_TLS_KEY_FILE]
   --tls_client_ca_file value      Path to the TLS CA file to verify clients against [$MICRO_TLS_CLIENT_CA_FILE]
   --api_address value             Set the api address e.g 0.0.0.0:8080 [$MICRO_API_ADDRESS]
   --proxy_address value           Proxy requests via the HTTP address specified [$MICRO_PROXY_ADDRESS]
   --web_address value             Set the web UI address e.g 0.0.0.0:8082 [$MICRO_WEB_ADDRESS]
   --network value                 Set the micro network name: local, go.micro [$MICRO_NETWORK]
   --network_address value         Set the micro network address e.g. :9093 [$MICRO_NETWORK_ADDRESS]
   --router_address value          Set the micro router address e.g. :8084 [$MICRO_ROUTER_ADDRESS]
   --gateway_address value         Set the micro default gateway address e.g. :9094 [$MICRO_GATEWAY_ADDRESS]
   --tunnel_address value          Set the micro tunnel address e.g. :8083 [$MICRO_TUNNEL_ADDRESS]
   --api_handler value             Specify the request handler to be used for mapping HTTP requests to services; {api, proxy, rpc} [$MICRO_API_HANDLER]
   --api_namespace value           Set the namespace used by the API e.g. com.example.api [$MICRO_API_NAMESPACE]
   --web_namespace value           Set the namespace used by the Web proxy e.g. com.example.web [$MICRO_WEB_NAMESPACE]
   --enable_stats                  Enable stats [$MICRO_ENABLE_STATS]
   --auto_update                   Enable automatic updates [$MICRO_AUTO_UPDATE]
   --report_usage                  Report usage statistics [$MICRO_REPORT_USAGE]
   --namespace value               Set the micro service namespace (default: "go.micro") [$MICRO_NAMESPACE]
   --plugin value                  Comma separated list of plugins e.g broker/rabbitmq, registry/etcd, micro/basic_auth, /path/to/plugin.so [$MICRO_PLUGIN]
   --help, -h                      show help
   --version, -v                   print the version
```
I'm assuming the nats plugin isn't in the build by default so some help on the best way to include it would be great, but I'm assuming this isn't the intended behavior.

**Comment by alex-godwin** [comment]
Alright looks like I can fix the issue by adding a `plugin.go` file with:

```go
package main

import (
	_ "github.com/micro/go-plugins/registry/nats"
)
```

and building micro again like `go build -o micro ./main.go ./plugin.go`. Is there a better way to load plugins than recompiling the binary? I saw references to `.so` files in the code when I was looking through, so I'm assuming there's support for using golang plugins, if so how would I load that at runtime?

**Comment by alex-godwin** [comment]
So I built the plugin using `micro plugin build --name registry/nats --type registry`, and that seems to start fine when running on my local machine with the command `./micro --registry=nats --plugin ./nats.so api`, but errors out in a docker container when I try to load it with the official `micro/micro` image with the error: `Error loading plugin ./nats.so: Failed to load plugin ./nats.so: plugin.Open("./nats.so"): Error loading shared library /nats.so: Exec format error`. My guess is there's an issue with the linker in alpine, but I'm stumped right now.

**Comment by afdalwahyu** [comment]
are you using macOS to build plugin? 
my guess is .so file you've built only works for your macOS machine. maybe you need to build plugin using linux machine.

anyway, is it good idea, if we can use prebuilt plugin (from ci/cd) and upload them, so we can use by using command like this @asim ?
```
$> micro --plugin https://plugin.micro.mu/v1.18.0/registry/kubernetes.so web
```

**Comment by asim** [comment]
Sorry I have not had a chance to look at this. I will do so now. Were you running NATS in a container also? 

Otherwise I think the idea of building the .so files using known releases and workflows make sense. I'll talk to @vtolstov about this.

**Comment by alex-godwin** [comment]
@afdalwahyu I think I must have accidentally done that instead of on my linux machine, I'm not running into the issue now, my bad. I forked the micro repo to test out building plugins when micro is compiled, just added this to the makefile:

```
plugins:
	mkdir -p ./soplugins
	./$(NAME) plugin build --name nats --type registry --output ./soplugins
```
and ran it after `micro` was built, but I ran into an issue that I can't replicate now, but I would get the error: 
```
Error loading plugin soplugins/nats.so: Failed to load plugin soplugins/nats.so: plugin.Open("soplugins/nats.so"): Error loading shared library /soplugins/nats.so:  plugin was built with a different version of package /some/package/
```
I think that the recent change to all plugins having their own go.mod files fixed that over the weekend, but worth mentioning.

Also @asim, I was running the NATS container in the same docker network it was an issue loading the plugin.
